crates/iota-rest-api/src/transactions/resolve.rs

```diff
@@ -0,0 +1,596 @@
+                matches_input_arg(split_coins.coin, arg_idx).then_some(None)
+            }
+            Command::MergeCoins(merge_coins) => {
+                if matches_input_arg(merge_coins.coin, arg_idx) {
+                    Some(None)
+                } else {
+                    merge_coins
+                        .coins_to_merge
+                        .iter()
+                        .position(|elem| matches_input_arg(*elem, arg_idx))
+                        .map(Some)
+                }
+            }
+            Command::Publish(_) => None,
+            Command::MakeMoveVector(make_move_vector) => make_move_vector
+                .elements
+                .iter()
+                .position(|elem| matches_input_arg(*elem, arg_idx))
+                .map(Some),
+            Command::Upgrade(upgrade) => matches_input_arg(upgrade.ticket, arg_idx).then_some(None),
+        }
+        .map(|x| (command, x))
+    })
+}
+
+fn matches_input_arg(arg: Argument, arg_idx: usize) -> bool {
+    matches!(arg, Argument::Input(idx) if idx as usize == arg_idx)
+}
+
+/// Estimate the gas budget using the gas_cost_summary from a previous DryRun
+///
+/// The estimated gas budget is computed as following:
+/// * the maximum between A and B, where: A = computation cost +
+///   GAS_SAFE_OVERHEAD * reference gas price B = computation cost + storage
+///   cost - storage rebate + GAS_SAFE_OVERHEAD * reference gas price overhead
+///
+/// This gas estimate is computed similarly as in the TypeScript SDK
+fn estimate_gas_budget_from_gas_cost(
+    gas_cost_summary: &GasCostSummary,
+    reference_gas_price: u64,
+) -> u64 {
+    const GAS_SAFE_OVERHEAD: u64 = 1000;
+
+    let safe_overhead = GAS_SAFE_OVERHEAD * reference_gas_price;
+    let computation_cost_with_overhead = gas_cost_summary.computation_cost + safe_overhead;
+
+    let gas_usage = gas_cost_summary.net_gas_usage() + safe_overhead as i64;
+    computation_cost_with_overhead.max(if gas_usage < 0 { 0 } else { gas_usage as u64 })
+}
+
+fn select_gas(
+    reader: &StateReader,
+    owner: IotaAddress,
+    budget: u64,
+    max_gas_payment_objects: u32,
+    input_objects: &[ObjectID],
+) -> Result<Vec<ObjectRef>> {
+    let gas_coins = reader
+        .inner()
+        .account_owned_objects_info_iter(owner, None)?
+        .filter(|info| info.type_.is_gas_coin())
+        .filter(|info| !input_objects.contains(&info.object_id))
+        .filter_map(|info| reader.inner().get_object(&info.object_id).ok().flatten())
+        .filter_map(|object| {
+            GasCoin::try_from(&object)
+                .ok()
+                .map(|coin| (object.compute_object_reference(), coin.value()))
+        })
+        .take(max_gas_payment_objects as usize);
+
+    let mut selected_gas = vec![];
+    let mut selected_gas_value = 0;
+
+    for (object_ref, value) in gas_coins {
+        selected_gas.push(object_ref);
+        selected_gas_value += value;
+    }
```

Comment:
Also not sure if that's intended to add all the gas coins here (to cleanup/merge objects maybe), but one could already break if `selected_gas_value` is `>= budget`

Reply:
Will be resolved in #7303